zstd.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/core/aligned_buffer.hh>
+
+// We need to use experimental features of the zstd library (to allocate compression/decompression context),
+// which are available only when the library is linked statically.
+#define ZSTD_STATIC_LINKING_ONLY
+#include <zstd.h>
+
+#include "compress.hh"
+#include "utils/class_registrator.hh"
+
+static const sstring COMPRESSION_LEVEL = "compression_level";
+static const sstring COMPRESSOR_NAME = compressor::namespace_prefix + "ZstdCompressor";
+
+class zstd_processor : public compressor {
+    int _compression_level = 3;
+
+    // Compression context.
+    std::unique_ptr<char[], free_deleter> _cctx_raw;
+    ZSTD_CCtx* _cctx;
+
+    // Decompression context.
+    std::unique_ptr<char[], free_deleter> _dctx_raw;
+    ZSTD_DCtx* _dctx;
+public:
+    zstd_processor(const opt_getter&);
+
+    size_t uncompress(const char* input, size_t input_len, char* output,
```

Comment:
Prefix overriden virtuals with `virtual`.

Reply:
That would be inconsistent with other compressors. See compress.cc.

Reply:
I'm looking but I'm seeing `virtual` prepended to such lines: https://github.com/scylladb/scylla/blob/7604980d638ef83cf8f8d8083fa1a82c00572ba4/sstables/compress.cc#L386

Yes, we have many undocumented de-facto coding standards that give way to pointless discussions around what is the *true* standard, as you can pretty much find precedence for anything in our code-base.

Reply:
https://github.com/scylladb/scylla/blob/7604980d638ef83cf8f8d8083fa1a82c00572ba4/compress.cc#L31

Reply:
Yeah, like I said, there can be found precedents for pretty-much any code-style.
BTW why was this not added to `compress.cc` like the other compressors?

Reply:
Separate compilation units = code can be compiled in parallel, which speeds up the compilation times. Plus it's more readable IMO (you don't spend time wondering which code belongs to which compressor). I'd even refactor what we have and put every compressor to a separate file in a directory like compressor/.

Reply:
To a certain point yes, but it also has a link-time overhead which hurts edit-compile-run cycles.
Also this is inconsistent with the other compressors. :)

Reply:
Linking is super fast though compared to compilation (especially if you use lld).
I can move the other compressors to separate files :P

Reply:
I'd keep the compressor in a separate file since it's not dependant on any of the other compressors. I don't think bundling multiple unrelated classes in a single file makes sense. The fact that they implement the same interface does not make them related.

Reply:
The problem with separate compilation units is that each compiles all its headers. This is less of a problem with compressors because they don't include so much (and on the other hand, they also don't benefit so buch because their dependencies rarely change).